ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/peers/MaintainedPeers.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.p2p.peers;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import tech.pegasys.pantheon.util.Subscribers;
+
+import java.util.Set;
+import java.util.stream.Stream;
+
+import io.vertx.core.impl.ConcurrentHashSet;
+
+/** Represents a set of peers for which connections should be actively maintained. */
+public class MaintainedPeers {
+  private final Set<Peer> maintainedPeers = new ConcurrentHashSet<>();
+  private final Subscribers<PeerAddedCallback> addedSubscribers = new Subscribers<>();
+  private final Subscribers<PeerRemovedCallback> removedCallbackSubscribers = new Subscribers<>();
+
+  public boolean add(final Peer peer) {
+    checkArgument(
+        peer.getEnodeURL().isListening(),
+        "Invalid enode url.  Enode url must contain a non-zero listening port.");
+    boolean wasAdded = maintainedPeers.add(peer);
+    addedSubscribers.forEach(s -> s.onPeerAdded(peer, wasAdded));
```

Comment:
Leaving this as-is for now because I don't want to get into changing the `Subscribers` interface.

Reply:
You don't have too.
I was just suggesting to do :    `addedSubscribers.parallelStream().forEach(s -> s.onPeerAdded(peer, wasAdded));` instead of `addedSubscribers.forEach(s -> s.onPeerAdded(peer, wasAdded));`